features/246-custom-domain/index.md

```diff
@@ -0,0 +1,70 @@
+# Custom domain
+
+## Overview
+
+Allow developer to setup custom domain for their application, basically following features will be supported.
+
+- Allow developer to setup custom domain.
+- Let's Encrypt certificate will be provided by default, certificate will be renew automatically.
+- Developer can configure to use custom SSL certificates for their domain. They will need to update the certificates before it expires in this case.
+
+## Use Cases
+
+### Setup custom domain through skycli
+
+1. Add custom domain
+
+    ```sh
+    $ skycli domain add myapp.example.com --app=myapp
+    Added domain myapp.example.com successfully!
+
+    Add following DNS records in your DNS provider.
+
+    TYPE      HOST                     VALUE
+    TXT       _skygear.example.com     5636486ffc5a4dfebf4a13f480bd9a95
```

Comment:
Not sure If I got it correct, but shouldn't it be `_skygear.myapp.example.com`?

Reply:
I did some research on some services that provide custom domain features:

### Firebase

When I added `a.example.com`, I need to add `TXT` record to `example.com` to verify the ownership. After I passed the verification, I can use any subdomains (e.g. `b.example.com`, `c.b.example.com`) to all of my apps. It seems that the verification is belong to user, if the subdomain is added by another collaborator, the `TXT` record verification step appear again.

### ZEIT

When I added `a.example.com`, I need to add `TXT` record to `_now.example.com` to verify the ownership. After adding `a.example.com`, we cannot use `example.com` in another account. User can add different subdomains in the same account.

So for us, I think that would be clearer for we to define the domain is belong to app. We used the TXT record verify developer own the domain. After developer verified the domain once, any new subdomains add to that app doesn't need to enter the TXT record again. So the host of the TXT record need to be under TLD+1. For example, if the TLD+1 is example.com, we will test _skygear.example.com. 

We can parse the domain with [EffectiveTLDPlusOne](https://godoc.org/golang.org/x/net/publicsuffix#EffectiveTLDPlusOne) function.

Thought? If it looks good, I can add a section to explain this.

Reply:
Am I correct that the verification process proves that the developer owns the root domain, and wants to assign the root domain to a specific app exclusively?

Do we support this situation?
- `forum.example.com` -> custom domain for `example-forum` app
- `blog.example.com` -> custom domain for `example-blog` app

Reply:
> Am I correct that the verification process proves that the developer owns the root domain, and wants to assign the root domain to a specific app exclusively?

yes

> Do we support this situation?
> 
> - forum.example.com -> custom domain for example-forum app
> - blog.example.com -> custom domain for example-blog app

Yes, but developers will need to add 2 different TXT records, as they are 2 different apps.

Reply:
Oh I see this has been discussed above. So the root domain assignment is not exclusive.